### PR TITLE
Add FFI exception translation

### PR DESCRIPTION
**Pull request by jlnr**
I've started with Gosu::Image, will tackle one .rb file per commit.

### CODE REVIEW

File: lib/gosu/channel.rb [hunk 1]
```diff
@@ -24,39 +24,46 @@ def __pointer
...
     def pan=(double)
-      _channel_set_pan(@__channel, double.clamp(0.0, 1.0))
+      Gosu_Channel_set_pan(@__channel, double.clamp(-1.0, 1.0))
+      Gosu.check_last_error
     end
-  end
 
-  def free_object
-    _destroy_channel(@__channel)
+    def free_object
```
Comment (jlnr) [inline lib/gosu/channel.rb]: @cyberarm This free_object method was defined on the Gosu module, not inside Gosu::Channel. I haven't had time to look at FFI memory management yet, not sure why Channel and Image have different approaches

File: lib/gosu.rb [hunk 2]
```diff
@@ -1,24 +1,36 @@
...
+    # Add this gem to the PATH on Windows so that bundled DLLs can be found.
+    # When running through Ocra on Windows, we need to be careful to preserve the ENV["PATH"]
+    # encoding (see #385).
+    path_encoding = ENV["PATH"].encoding
+    ENV["PATH"] = "#{binary_path.encode(path_encoding)};#{ENV["PATH"]}"
+  end
 
+  # Add the correct lib directory for the current version of Ruby (major.minor).
+  $LOAD_PATH.unshift File.join(binary_path, RUBY_VERSION[/^\d+.\d+/])
+end
```
Comment (jlnr) [inline lib/gosu.rb:24]: Moved this into the main `lib/gosu.rb` file because that it seemed like the more obvious place.

File: lib/gosu.rb [hunk 3]
```diff
@@ -245,4 +259,28 @@ def self.blend_mode(mode)
...
   end
+
+  def self.check_last_error
+    if (err = Gosu_last_error())
+      raise err
+    end
+  end
 end
+
+# Individual classes need to be loaded after defining Gosu.check_last_error.
```
Comment (jlnr) [inline lib/gosu.rb:344]: :/

Gosu::Color::RED calls Gosu::Color#initialize, which calls Gosu.check_last_error. Not sure if there is a better way around this.

**Comment by jlnr** [comment]
Done & ready for review ✌️

File: lib/gosu/channel.rb (hunk 1 above)
Comment (cyberarm) [reply lib/gosu/channel.rb]: Hmm, I think that was a hold over from when I was figuring out how FFI worked. I believe the "correct" way would be like it's done in Font or Image where you create an `FFI::AutoPointer` and give it a method to call when it's destroyed/gc'd.
https://github.com/cyberarm/ffi-gosu/blob/cf21b6714227c6992017e60a6ba8719c61c5922c/lib/gosu/font.rb#L28-L35
https://github.com/cyberarm/ffi-gosu/blob/cf21b6714227c6992017e60a6ba8719c61c5922c/lib/gosu/font.rb#L90-L92

**Review by cyberarm — CHANGES_REQUESTED** [review]

File: lib/gosu.rb [hunk 4]
```diff
@@ -1,82 +1,111 @@
...
-require_relative "gosu/color"
-require_relative "gosu/text_input"
-require_relative "gosu/gl_tex_info"
-require_relative "gosu/channel"
-require_relative "gosu/sample"
-require_relative "gosu/song"
-
-require_relative "gosu/compat"
+if RUBY_PLATFORM =~ /mswin$|mingw32|mingw64|win32\-|\-win32/
+  binary_path = File.expand_path("../../../gosu/lib", __dir__)#File.dirname(__FILE__)
```
Comment (cyberarm) [inline lib/gosu.rb:5]: This hard coded path needs a `../` removed since it's been moved "up" a directory.

File: lib/gosu.rb [hunk 5]
```diff
@@ -92,94 +121,153 @@ def self.gl(z = nil, &block)
...
+  def self.distance(x1, y1, x2, y2)
+    Gosu_distance(x1, y1, x2, y2).tap { Gosu.check_last_error }
+  end
+
+  def self.angle(x1, y1, x2, y2)
+    Gosu_angle(x1, y1, x2, y2).tap { Gosu.check_last_error }
+  end
+
+  def self.angle_diff(angle1, angle2)
+    Gosu_angle_diff(angle, angle2).tap { Gosu.check_last_error }
```
Comment (cyberarm) [inline lib/gosu.rb]: `angle` should be `angle1` here: `Gosu_angle_diff(angle1, angle2)`

**Review by cyberarm — CHANGES_REQUESTED** [review]

File: lib/gosu/channel.rb (hunk 1 above)
Comment (jlnr) [reply lib/gosu/channel.rb]: Sounds good.

**Comment by jlnr** [comment]
Thanks for the thorough review. I've made all changes, tests are green. If you have a game that's ready to be tested with this branch, that'd be great – the tests only cover so much of Gosu.

File: lib/gosu/image.rb [hunk 6]
```diff
@@ -100,37 +108,41 @@ def __pointer
...
     end
 
     def to_blob
-      _image_to_blob(__pointer).read_string(width * height * Gosu::Color::SIZEOF)
+      Gosu_Image_to_blob(__pointer).tap { Gosu.check_last_error }.read_string(width * height * Gosu::Color::SIZEOF)
     end
 
     def subimage(left, top, width, height)
-      Gosu::Image.new( _image_subimage(__pointer, left, top, width, height) )
+      Gosu::Image.new(Gosu_Image_subimage(__pointer, left, top, width, height).tap { Gosu.check_last_error })
```
Comment (cyberarm) [inline lib/gosu/image.rb]: `Gosu_Image_subimage` should be `Gosu_Image_create_from_subimage` here.

File: lib/gosu/image.rb [hunk 7]
```diff
@@ -56,42 +64,42 @@ def self.load_tiles(filename_or_image, tile_width ,tile_height, retro: false, ti
       callback = proc { |data, image| images << Gosu::Image.new(image, retro: retro, tileable: tileable) }
 
       if filename_or_image.is_a?(String)
-        _image_load_tiles(filename_or_image, tile_width, tile_height, callback, nil, flags)
+        Gosu_Image_load_tiles(filename_or_image, tile_width, tile_height, callback, nil, flags)
```
Comment (cyberarm) [inline lib/gosu/image.rb]: `Gosu_Image_load_tiles` should be `Gosu_Image_create_from_tiles`

File: lib/gosu/image.rb [hunk 8]
```diff
@@ -56,42 +64,42 @@ def self.load_tiles(filename_or_image, tile_width ,tile_height, retro: false, ti
...
       else
         if filename_or_image.instance_of?(Gosu::Image)
-          _image_load_tiles_from_image(filename_or_image.__pointer, tile_width, tile_height, callback, nil, flags)
+          Gosu_Image_load_tiles_from_image(filename_or_image.__pointer, tile_width, tile_height, callback, nil, flags)
         else
-          _image_load_tiles_from_image(Gosu::Image.new(filename_or_image, retro: retro, tileable: tileable).__pointer, tile_width, tile_height, callback, nil, flags)
+          Gosu_Image_load_tiles_from_image(Gosu::Image.new(filename_or_image, retro: retro, tileable: tileable).__pointer, tile_width, tile_height, callback, nil, flags)
```
Comment (cyberarm) [inline lib/gosu/image.rb]: Same here

File: lib/gosu/image.rb [hunk 9]
```diff
@@ -56,42 +64,42 @@ def self.load_tiles(filename_or_image, tile_width ,tile_height, retro: false, ti
       callback = proc { |data, image| images << Gosu::Image.new(image, retro: retro, tileable: tileable) }
 
       if filename_or_image.is_a?(String)
-        _image_load_tiles(filename_or_image, tile_width, tile_height, callback, nil, flags)
+        Gosu_Image_load_tiles(filename_or_image, tile_width, tile_height, callback, nil, flags)
       else
         if filename_or_image.instance_of?(Gosu::Image)
-          _image_load_tiles_from_image(filename_or_image.__pointer, tile_width, tile_height, callback, nil, flags)
+          Gosu_Image_load_tiles_from_image(filename_or_image.__pointer, tile_width, tile_height, callback, nil, flags)
```
Comment (cyberarm) [inline lib/gosu/image.rb]: `Gosu_Image_load_tiles_from_image` should be `Gosu_Image_create_tiles_from_image`

**Review by cyberarm — CHANGES_REQUESTED** [review]
Looks like I had renamed a few `Image_create` methods that messed up your renaming.

**Comment by cyberarm** [comment]
Tried to run I-MIC RTS, crashed pretty quick when slicing the tiles 🙂 🗺️

**Comment by jlnr** [comment]
I've fixed the image method calls.

**Comment by cyberarm** [comment]
Merged! Thanks for getting this done! 🎉

**Pull request merged**